### PR TITLE
Add table of contents to constants.md

### PR DESCRIPTION
**Pull request by vietlubu**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

[//]: # (Describe at length, the changes that this pull request makes.)

- Change file `src/plugins/constdb2doc.c` to generate TOC for `doc/constants.md`
- Update file `doc/constants.md` with TOC (by `constdb2doc` plugin)

**Affected Branches:** 

[//]: # (Master? Slave?)
`master `

**Issues addressed:**

[//]: # (Issue Tracker Number if any.)

### Known Issues and TODO List

[//]: # (Insert checklist here)
[//]: # (Syntax: - [ ] Checkbox)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1996)

**Comment by MishimaHaruna** [comment]
Thank you for the idea, but please note that the file you're editing is an automatically generated file, and it will be overwritten at the next commit, as mentioned at the end of the file:

```
<!--GENERATED FILE DO NOT EDIT-->
```

In order to make the change permanent, it needs to be implemented in the generator plugin instead (see  src/plugins/constdb2doc.c for more information)

**Comment by vietlubu** [comment]
Sorry, my mistake. 😂 I will try to update `src/plugins/constdb2doc.c` 💭

File: src/plugins/constdb2doc.c [hunk 1]
```diff
@@ -180,7 +229,28 @@ void do_constdb2doc(void)
 	fprintf(out_fp, "> End of list\n\n");
 	fprintf(out_fp, "<!--GENERATED FILE DO NOT EDIT-->\n");
 
+	fclose(out_header);
 	fclose(out_fp);
+
+
+	// char * line = NULL;
```
Comment (4144) [inline src/plugins/constdb2doc.c]: please not add commented code

File: src/plugins/constdb2doc.c [hunk 2]
```diff
@@ -180,7 +229,28 @@ void do_constdb2doc(void)
...
+    size_t len = 255;
+    char *line = malloc(sizeof(char) * len);
+
+	while (fgets(line, len, out_header) != NULL)  {
+		printf(line);
+	}
+	free(line);
+	// if(line) {
```
Comment (4144) [inline src/plugins/constdb2doc.c]: and this code

File: src/plugins/constdb2doc.c [hunk 3]
```diff
@@ -180,7 +229,28 @@ void do_constdb2doc(void)
...
+	// 	free(line);
+	// }
+
+	fclose(out_header);
+
+
```
Comment (4144) [inline src/plugins/constdb2doc.c]: probably remove extra empty lines here

File: src/plugins/constdb2doc.c [hunk 4]
```diff
@@ -47,17 +48,50 @@ HPExport struct hplugin_info pinfo = {
...
 };
 
+FILE *out_header;
 FILE *out_fp;
 bool torun = false;
+char *anchor;
+
+char * get_anchorid(char anchorname[])
+{
+	// anchorid = anchorname;
```
Comment (4144) [inline src/plugins/constdb2doc.c]: please not add commented code

**Comment by vietlubu** [comment]
@4144 Thank for your review, pull request in `[WIP]`.
Current, I want to make 2 files. One for header (TOC), one for contents. I have the issue when do merge 2 files. Do you have any ideas for this step? Thank you!

**Comment by 4144** [comment]
simplest way what i can think.
1. create two files by current code. save toc into constants.md and other file with any temp name.
2. open for read temp file with constants and read it.
3. open for append constants.md and write here data from step 2.

**Comment by vietlubu** [comment]
@4144 Please review if you have free time. Thank you!

**Comment by 4144** [comment]
@vietlubu can you squash your commits?

File: src/plugins/constdb2doc.c [hunk 5]
```diff
@@ -47,17 +48,53 @@ HPExport struct hplugin_info pinfo = {
...
+	char *anchor = (char *) malloc(256);
+
+	strcpy(anchor, anchorstring);
+
+	int i, j;
+	for (i = 0, j = 0; anchor[i] != 0; i++) {
+		anchor[i] = tolower(anchor[i]);
+
+		// Check space
+		if (anchor[i] == 32) {
```
Comment (4144) [inline src/plugins/constdb2doc.c]: better use chars and not decimal constants.
32 -> ' '
45 -> '-'
etc

File: src/plugins/constdb2doc.c [hunk 6]
```diff
@@ -157,17 +204,45 @@ void constdb2doc_itemdb(void)
 	fprintf(out_fp, "\n");
 }
 
+void merge_files()
```
Comment (4144) [inline src/plugins/constdb2doc.c]: better read file line by line or whole file at once.

for read line by line, open file with mode "rt" and read by fgets, and write by fputs

for read whole file, use ftell to get size, allocated buffer with given size, then use fread for read whole file, and fwrite for write all at once.

Comment (vietlubu) [reply src/plugins/constdb2doc.c]: Thank you, for the solution. I updated.

**Comment by vietlubu** [comment]
@4144 Please review this PR if you have free time. Thank you! ✌️

**Comment by 4144** [comment]
look like you still using ints for chars. i remember I saw fixed version already

File: src/plugins/constdb2doc.c [hunk 7]
```diff
@@ -215,10 +215,16 @@ void merge_files()
...
+	long int len;
+	fseek(out_fp, 0, SEEK_END);
+	len = ftell(out_fp);
+
+	char buffer[len];
+
+	fseek(out_fp, 0, SEEK_SET);
+	fread(buffer, len, 1, out_fp);
+
+	fwrite(buffer , 1 , sizeof(buffer) , out_header);
```
Comment (4144) [inline src/plugins/constdb2doc.c]: change to ``fwrite(buffer, 1, sizeof(buffer), out_header);``

**Comment by 4144** [comment]
also please squash two commits into one

**Pull request closed without merging**